### ISSUE TITLE
Using scale and rotate augmentations together can result in inaccurate bounding boxes.

### ISSUE THREAD

**Issue by neobaud** [ultralytics/yolov5#6138]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training

### Bug

Using scale and rotate augmentations together can result in inaccurate bounding boxes.  This seems to be due to the box being scaled and rotated in one operation, and then clipped afterwards.

![train_batch0](https://user-images.githubusercontent.com/66406852/147748782-5e61ec88-8d3a-4785-adfa-f52608ce07b0.jpg)
![train_batch1](https://user-images.githubusercontent.com/66406852/147748784-23af9531-94eb-4f00-8a97-cb5d02dc5542.jpg)
![train_batch2](https://user-images.githubusercontent.com/66406852/147748786-b06a9db8-4c4c-43a8-a5b8-a5cff5bb2a13.jpg)

### Environment

- YOLO : YOLOv5 🚀 v6.0-159-gdb6ec66 torch 1.9.1 CUDA:0 (NVIDIA GeForce RTX 3090, 24260MiB)
- OS : Ubuntu 20.04
- Python : 3.8.8

### Minimal Reproducible Example

Dataset : I extracted just the class "trains" from coco and assembled the data in /data/cocotrains/dataset.yaml

Training command 
```
python3 train.py --batch -1 --cfg models/yolov5s.yaml --hyp data/hyps/hyp.cust.yaml --data /data/cocotrains/dataset.yaml --weights yolov5s.pt
```

Hyper-parameters (contents of hyp.cust.yaml): 
```
# YOLOv5 🚀 by Ultralytics, GPL-3.0 license
# Hyperparameters for COCO training from scratch
# python train.py --batch 40 --cfg yolov5m.yaml --weights '' --data coco.yaml --img 640 --epochs 300
# See tutorials for hyperparameter evolution https://github.com/ultralytics/yolov5#tutorials

lr0: 0.01  # initial learning rate (SGD=1E-2, Adam=1E-3)
lrf: 0.1  # final OneCycleLR learning rate (lr0 * lrf)
momentum: 0.937  # SGD momentum/Adam beta1
weight_decay: 0.0005  # optimizer weight decay 5e-4
warmup_epochs: 3.0  # warmup epochs (fractions ok)
warmup_momentum: 0.8  # warmup initial momentum
warmup_bias_lr: 0.1  # warmup initial bias lr
box: 0.05  # box loss gain
cls: 0.5  # cls loss gain
cls_pw: 1.0  # cls BCELoss positive_weight
obj: 1.0  # obj loss gain (scale with pixels)
obj_pw: 1.0  # obj BCELoss positive_weight
iou_t: 0.20  # IoU training threshold
anchor_t: 4.0  # anchor-multiple threshold
# anchors: 3  # anchors per output layer (0 to ignore)
fl_gamma: 0.0  # focal loss gamma (efficientDet default gamma=1.5)
hsv_h: 0.015  # image HSV-Hue augmentation (fraction)
hsv_s: 0.7  # image HSV-Saturation augmentation (fraction)
hsv_v: 0.4  # image HSV-Value augmentation (fraction)
degrees: 30.0  # image rotation (+/- deg)
translate: 0.1  # image translation (+/- fraction)
scale: 0.9  # image scale (+/- gain)
shear: 0.0  # image shear (+/- deg)
perspective: 0.0  # image perspective (+/- fraction), range 0-0.001
flipud: 0.0  # image flip up-down (probability)
fliplr: 0.5  # image flip left-right (probability)
mosaic: 1.0  # image mosaic (probability)
mixup: 0.0  # image mixup (probability)
copy_paste: 0.0  # segment copy-paste (probability)
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@neobaud the augmented labels enclose the augmented box perfectly. If you rotate a square 45 degrees and enclose it in a new square this is the natural result. There is no bug.

**Comment by neobaud** [comment]
Take the bottom row, 2nd image from the right.

Both trains in these images have very loose bounding boxes.  

If you took the four points from the annotation bounding boxes, rotated them and scaled them, clipped them to the new view plane then enclosed them in a new bounding box, the resulting box would be tighter around the train than is evident here.

**Comment by glenn-jocher** [comment]
@neobaud the important point is that the augmented labels don't enclose augmented trains, they enclose the augmented original labels, which are just squares. 

If you want to minimize rotation effects on labels just keep your rotation angles low.

**Comment by neobaud** [comment]
In order to correctly rotate, scale and clip the bounding boxes, you need to deal with the individual points of the bounding boxes, rather than the boxes as a whole.

Are you interested in a fix?

**Comment by glenn-jocher** [comment]
@neobaud the only solution is to use segmentation labels. This is already implemented, so if you use YOLOv5 segmentation labels, they will be detected and handled automatically. Segmentation format is segment arrays (normalized):
```
class x1, y1, x2, y2, x3, y3, ..., xn, yn
```

Each line can have any number of points. See https://github.com/ultralytics/yolov5/releases/download/v1.0/coco128-segments.zip

**Comment by glenn-jocher** [comment]
@neobaud there's another way to deal with this also for normal labels which is just to **estimate** a certain amount of clipping as a function of rotation angle, i.e. so if you rotate 45 deg you may assume statistically that you should shrink your box by a certain scale factor. You could probably come up some empirical scaling curve based on the COCO segment vs box labels, but this is significant work, and may not generalize well to other datasets. Just an idea. 

There are some examples out there of users discovering the same issue:
![image](https://user-images.githubusercontent.com/26833433/147832797-0294eb09-b409-4db4-8255-2842e925398d.png)
